src/server/semester/semester.service.ts

```diff
@@ -59,4 +62,150 @@ export class SemesterService {
 
       );
   }
+
+  /**
+   * Creates a new academic year of course instances, non class events, and
+   * faculty absences using the values of the previous semester.
+   * Retired course instances retain their "retired" status. Meetings and
+   * enrollment values are not retained in the newly created academic year.
+   */
+  public async addAcademicYear(newAcademicYear: number): Promise<void> {
+    // Note that academicYear in the semester table is actually calendar year.
+    const existingYears = await this.getYearList();
+    // If the academic year previous to the newAcademicYear does not exist,
+    // throw an exception.
+    if (!existingYears.includes((newAcademicYear - 1).toString())) {
+      throw new BadRequestException(
+        'Cannot create requested academic year until preceding academic year is created.'
+      );
+    }
+    // Only create the new academic year if it does not yet exist.
+    if (!existingYears.includes((newAcademicYear).toString())) {
+      // Try getting the year previous to the new academic year requested.
+      // For example, if we are trying to create the 2026 academic year, use
+      // the semesters Spring 2025 and Fall 2024 to construct the new year.
+      const existingFallSemester = await this.semesterRepository
+        .findOneOrFail({
+          where: {
+            // NOTE: academic year in database is actually calendar year
+            term: TERM.FALL,
+            academicYear: newAcademicYear - 2,
+          },
+          // See: https://github.com/typeorm/typeorm/issues/1270#issuecomment-348429760
+          relations: ['absences', 'courseInstances', 'courseInstances.course', 'nonClassEvents'],
+        });
+
+      // Get the most recent existing spring semester along with its relations
+      const existingSpringSemester = await this.semesterRepository
+        .findOneOrFail({
+          where: {
+            // NOTE: academic year in database is actually the calendar year
+            term: TERM.SPRING,
+            academicYear: newAcademicYear - 1,
+          },
+          // See: https://github.com/typeorm/typeorm/issues/1270#issuecomment-348429760
+          relations: ['absences', 'courseInstances', 'courseInstances.course', 'nonClassEvents'],
+        });
```

Comment:
These `findOneOrFail` calls should probably be wrapped in a `try ... catch` to handle the case where an existing semester can't be found.

Reply:
Good point. The calls to find the semester were updated so that they're wrapped in a `try...catch` in [bc9c831](https://github.com/seas-computing/course-planner/pull/496/commits/bc9c831e0468e0a628af5408b6cbe272ab9f7dcd).